addon/components/flash-message.js

```diff
@@ -2,127 +2,110 @@ import { htmlSafe, classify } from '@ember/string';
-    this.element.removeEventListener('mouseleave', this._mouseLeaveHandler);
-  },
-
-  progressDuration: computed('flash.showProgress', {
-    get() {
-      if (!get(this, 'flash.showProgress')) {
-        return false;
-      }
-
-      const duration = getWithDefault(this, 'flash.timeout', 0);
-
-      return htmlSafe(`transition-duration: ${duration}ms`);
+  @readOnly('flash.showProgress')
+  showProgress;
+
+  @not('exiting')
+  notExiting;
+
+  @and('showProgress', 'notExiting')
+  showProgressBar;
+
+  @readOnly('flash.exiting')
+  exiting;
+
+  @bool('template')
+  hasBlock;
+
+  @computed('flash.type', 'messageStyle')
+  get alertType() {
+    const flashType = this.flash.type || '';
+    const messageStyle = this.messageStyle || '';
+    let prefix = 'alert alert-';
+
+    if (messageStyle === 'foundation') {
+      prefix = 'alert-box ';
     }
-  }),
 
-  click() {
-    const destroyOnClick = getWithDefault(this, 'flash.destroyOnClick', true);
+    return `${prefix}${flashType}`;
+  }
 
-    if (destroyOnClick) {
-      this._destroyFlashMessage();
+  @computed('flash.type')
+  get flashType() {
+    return classify(this.flash.type || '');
+  }
+
+  @computed('flash.{showProgress,timeout}')
+  get progressDuration() {
+    if (!this.flash?.showProgress) {
+      return false;
     }
-  },
+    const duration = this.flash?.timeout || 0;
+    return htmlSafe(`transition-duration: ${duration}ms`);
+  }
 
   _mouseEnter() {
-    const flash = get(this, 'flash');
-    if (isPresent(flash)) {
-      flash.preventExit();
+    if (isPresent(this.flash)) {
+      this.flash.preventExit();
     }
-  },
+  }
 
   _mouseLeave() {
-    const flash = get(this, 'flash');
-    if (isPresent(flash) && !get(flash, 'exiting')) {
-      flash.allowExit();
+    if (isPresent(this.flash) && !this.flash.exiting) {
+      this.flash.allowExit();
     }
-  },
-
-  willDestroy() {
-    this._super(...arguments);
-    this._destroyFlashMessage();
-    cancel(get(this, 'pendingSet'));
-  },
+  }
 
-  // private
   _destroyFlashMessage() {
-    const flash = getWithDefault(this, 'flash', false);
-
-    if (flash) {
-      flash.destroyMessage();
+    if (this.flash) {
+      this.flash.destroyMessage();
     }
-  },
+  }
 
-  actions: {
-    close() {
+  @action
+  onClick() {
+    const destroyOnClick = this.flash?.destroyOnClick ?? true;
+
+    if (destroyOnClick) {
       this._destroyFlashMessage();
     }
   }
-});
+
+  @action
+  onClose() {
+    this._destroyFlashMessage();
+  }
+
+  @action
+  onDidInsert(element) {
+    const pendingSet = next(this, () => {
+      set(this, 'active', true);
+    });
+    set(this, 'pendingSet', pendingSet);
+    set(this, '_mouseEnterHandler', this._mouseEnter);
+    set(this, '_mouseLeaveHandler', this._mouseLeave);
+    element.addEventListener('mouseenter', this._mouseEnterHandler);
+    element.addEventListener('mouseleave', this._mouseLeaveHandler);
```

Comment:
I'm not positive, but I think you might need the `@action` decorator on `this._mouseEnter` and `this._mouseLeave` to preserve the context. Other than that this all looks great!

Reply:
Since the `@action` decorator is used on `onDidInsert` the references in its body will be bound to the component context, and since the handler functions are internal and not mutable (shouldn't be anyway) I don't think we should have an issue leaving them as "private" functions without the action decorator. I could be mistaken though, nothing is ever straightforward when dealing with `this`. The dummy app seems to behave as expected though and all tests pass.